ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,41 @@
+# Choosing a Versioning Schemes
+
+This document records our discussions around choosing how to version our packages.
+
+## Desiderata
+
+- *Simplicity and Familiarity*. We'd like the versioning scheme to be simple to explain, and ideally already well-established.
+
+- *Ease of execution*. We'd like the process of cutting a release to merely involve following a very simple checklist. We'd like it to require as few inputs, discussions, decisions, etc.
+
+- *Distinguished development versions*. We'd like to distinguish between the two possible semantics of version number.
+
+    - The version of a released thing identifies some immutable thing, which is always somehow more refined than any release that has a lesser version number.
+
+    - A development version refers to some _mutable_ thing that is improving during the time between two releases, eg the version on master. IE there will usually be multiple different commits that all have the same development version number.
+
+## Proposal Simplest
+
+To cut a release, merely create branch release/XXX pointing at the desired commit on master and also immediately create a subsequent master commit that advances all the appropriate versions.
```

Comment:
> create branch release/XXX

Just to be sure: In all proposals listed here, cutting a release involves "create branch release/XXX", i.e. you would create a new branch for every version -- is this intentional?

E.g. right now, we just have *one* release branch ([`release/cardano-node-1.35.x`](https://github.com/input-output-hk/ouroboros-network/tree/release/cardano-node-1.35.x)), which was forked off of `master`, and only contains backported commits and version bumps, which are referenced in CHaP.

Reply:
I'd suggest using tags, instead of branches for releases.

Reply:
I am also in favour of `tags`, but branches might be needed at times so I wouldn't rule them out.  This is because the networking team might push changes to master which we might not want to release in a hot fix release, but rather a major release; or the other way around: like currently with `1.35.5` where I needed to push >100 commits to the release branch to avoid releasing consensus changes.  *That's actually another argument to split network & consensus into two repos (I am starting to convince myself that's actually a good thing to do).*

If we had two repositories, we would integrate network changes to master, but then to try it out, we would likely cherry pick changes on the most recent `ouroboros-consensus` released with `cardano-node`.  Since now only `ouroboros-consensus-diffusion` depends on `ouroboros-network`, merging our changes to `master` might be enough to avoid using not yet released `ouroboros-consensus` with our new stuff. If we add a new version of `NodeToNodeVersion` (or `NodeToClientVersion`) then we will also need a newer version of `ouroboros-consensus-cardano` for which we might need to use a release branch.

Reply:
> This is because the networking team might push changes to master which we might not want to release in a hot fix release, but rather a major release; or the other way around: 

In such situation you can always create a branch from a tag, work on it and once such fix has finalized and release was made it should be turned into another tag. Branches are intended for mutable things, while tags for immutable, such as releases.

Reply:
Do you want to release all the consensus packages at the same time, or only release packages which have some changes? Probably the former, but it's better to be explicit about it.

Reply:
Sorry, yes, I wrote too loosely. I should have consistently said:

> To cut a release of a branch new minor version, create a branch release/foo.xx and ...

I think that's our plan (and we'll tag increments within that branch with concrete verssions `release/foo.1.37.3`).

Reply:
:slightly_smiling_face: perhaps `release/foo-A.B.3`

Reply:
I cleaned it up to just say "announce a commit" instead of talking about branches, tags, etc. Thanks!